ArduCopter/events.cpp

```diff
@@ -4,152 +4,235 @@
+            g2.afs.gcs_terminate(true, battery_type_str);
 #else
-                arming.disarm();
+            arming.disarm();
 #endif
-        }
     }
 }
 
 // failsafe_gcs_check - check for ground station failsafe
 void Copter::failsafe_gcs_check()
 {
-    if (failsafe.gcs) {
-        // we must run the failsafe checks if we are in failsafe -
-        // otherwise we will never leave failsafe
-    } else if (g.failsafe_gcs == FS_GCS_DISABLED) {
-        // simply disabled
-        return;
-    } else if (failsafe.last_heartbeat_ms == 0) {
-        // GCS has never connected
-        return;
-    } else if (RC_Channels::has_active_overrides()) {
-        // GCS is currently telling us what to do!
-    } else if (control_mode == Mode::Number::GUIDED ||
-               control_mode == Mode::Number::GUIDED_NOGPS) {
-        // GCS is currently telling us what to do!
-    } else {
+    // Bypass GCS failsafe checks if disabled or GCS never connected
+    if (g.failsafe_gcs == FS_GCS_DISABLED || failsafe.last_heartbeat_ms == 0) {
         return;
     }
 
     // calc time since last gcs update
     // note: this only looks at the heartbeat from the device id set by g.sysid_my_gcs
     const uint32_t last_gcs_update_ms = millis() - failsafe.last_heartbeat_ms;
 
-    // check if all is well
-    if (last_gcs_update_ms < FS_GCS_TIMEOUT_MS) {
-        // check for recovery from gcs failsafe
-        if (failsafe.gcs) {
-            failsafe_gcs_off_event();
-            set_failsafe_gcs(false);
-        }
-        return;
-    }
+    // Determine which event to trigger
+    if (last_gcs_update_ms < FS_GCS_TIMEOUT_MS && failsafe.gcs) {
+        // Recovery from a GCS failsafe
+        set_failsafe_gcs(false);
+        failsafe_gcs_off_event();
 
-    // do nothing if gcs failsafe already triggered or motors disarmed
-    if (failsafe.gcs || !motors->armed()) {
-        return;
+    } else if (last_gcs_update_ms < FS_GCS_TIMEOUT_MS && !failsafe.gcs) {
+        // No problem, do nothing
+
+    } else if (last_gcs_update_ms > FS_GCS_TIMEOUT_MS && failsafe.gcs) {
+        // Already in failsafe, do nothing
+
+    } else if (last_gcs_update_ms > FS_GCS_TIMEOUT_MS && !failsafe.gcs) {
+        // New GCS failsafe event, trigger events
+        set_failsafe_gcs(true);
+        failsafe_gcs_on_event();
     }
+}
 
-    // GCS failsafe event has occurred
-    set_failsafe_gcs(true);
+// failsafe_gcs_on_event - actions to take when GCS contact is lost
+void Copter::failsafe_gcs_on_event(void)
+{
     AP::logger().Write_Error(LogErrorSubsystem::FAILSAFE_GCS, LogErrorCode::FAILSAFE_OCCURRED);
-
-    // clear overrides so that RC control can be regained with radio.
     RC_Channels::clear_overrides();
 
-    if (should_disarm_on_failsafe()) {
+    int16_t action;
+    action = g.failsafe_gcs; //The action set in FS_GCS_ENABLE
+
+    // Conditions to deviate from FS_GCS_ENABLE parameter setting
+    if (!motors->armed()) {
+        action = FS_GCS_DISABLED;
```

Comment:
The battery failsafe does something very similar but it uses the Failsafe_Action enum.  We could consolidate on the battery failsafe method by adding another switch.  what do you think?

Reply:
Not sure what you mean?